### PR TITLE
Use goreleaser to build releases

### PR DESCRIPTION
**Pull request by WillAbides**
This speeds up `make bufrelease` on my laptop from 6+ minutes to 20
seconds.

Instead of running go build in a docker container for every binary it
builds, `make bufrelease` now uses `goreleaser release` to build release
assets.

To ensure `go build` uses the correct go version, it:
- Sets RELEASE_GO_BINARY=go${GO_VERSION}
- Downloads with `go install
  "golang.org/dl/${RELEASE_GO_BINARY}"@latest` and `${RELEASE_GO_BINARY}
download`
- Symlinks `.build/release/go` to `${GOBIN}/${RELEASE_GO_BINARY}`
- Sets the gobinary config in goreleaser to `.build/release/buf/go`

A new file `data/goversion` stores the go version used for builds. This
is just because it's slightly easier to set there vs using set to update
a .mk file.

~~The new version makes the same assets as the old version. It also
creates `buf-Windows-x86_64.zip` and `buf-Windows-arm64.zip`.~~
It no longer creates zip files for windows.

### CODE REVIEW

**Comment by WillAbides** [comment]
@bufdev I added `go mod verify` to the release script so it will now fail if any dependencies have been modified in the module cache.

File: make/buf/scripts/release.bash [hunk 1]
```diff
@@ -11,137 +11,42 @@ fail() {
...
-  # minisign installed.
-  secret_key_file="$(mktemp)"
-  trap "rm ${secret_key_file}" EXIT
-  # Prevent printing of private key and password
-  set +x
-  echo "${RELEASE_MINISIGN_PRIVATE_KEY}" > "${secret_key_file}"
-  echo "${RELEASE_MINISIGN_PRIVATE_KEY_PASSWORD}" | minisign -S -s "${secret_key_file}" -m .build/release/buf/assets/sha256.txt
-  set -x
-  exit 0
+if [ -n "${RELEASE_SKIP_SIGN}" ]; then
```
Comment (bufdev) [inline make/buf/scripts/release.bash:16]: These variables need to be commented on as to what they do

File: make/buf/scripts/release.bash [hunk 2]
```diff
@@ -11,137 +11,42 @@ fail() {
...
+  SKIP_SIGN="--skip-sign"
+elif [ -z "${RELEASE_MINISIGN_PRIVATE_KEY}" -o -z "${RELEASE_MINISIGN_PRIVATE_KEY_PASSWORD}" ]; then
+  fail "RELEASE_MINISIGN_PRIVATE_KEY and RELEASE_MINISIGN_PRIVATE_KEY_PASSWORD must be set."
 fi
 
-BASE_NAME="buf"
+[ -z "${RELEASE_SKIP_VALIDATE}" ] || SKIP_VALIDATE="--skip-validate"
```
Comment (bufdev) [inline make/buf/scripts/release.bash:22]: We use `if [ -z ... ]; then`

File: make/buf/scripts/release.bash [hunk 3]
```diff
@@ -11,137 +11,42 @@ fail() {
...
+ASSETS_DIR="${RELEASE_DIR}/assets"
+WORKSPACE_DIR="${RELEASE_DIR}/workspace"
+DIST_DIR="${WORKSPACE_DIR}/dist"
 
-RELEASE_DIR=".build/release/${BASE_NAME}"
 rm -rf "${RELEASE_DIR}"
-mkdir -p "${RELEASE_DIR}"
-cd "${RELEASE_DIR}"
+mkdir -p "${ASSETS_DIR}" "${WORKSPACE_DIR}"
+trap "rm -rf $WORKSPACE_DIR" EXIT
```
Comment (bufdev) [inline make/buf/scripts/release.bash:31]: Needs `{}` - see the bash style guide, this is true for this whole file

File: make/buf/scripts/release.bash [hunk 4]
```diff
@@ -11,137 +11,42 @@ fail() {
...
+
+test -f "${GOBIN}/${RELEASE_GO_VERSION}" || go install "golang.org/dl/${RELEASE_GO_VERSION}"@latest
```
Comment (bufdev) [inline make/buf/scripts/release.bash:33]: `if`

Comment (bufdev) [reply make/buf/scripts/release.bash:33]: Also include quote with `@latest`

File: make/buf/scripts/release.bash [hunk 5]
```diff
@@ -11,137 +11,42 @@ fail() {
...
+"${GOBIN}/${RELEASE_GO_VERSION}" download 2>/dev/null
```
Comment (bufdev) [inline make/buf/scripts/release.bash:34]: Usage of `GOBIN` in this script scares me, this is the project `$GOBIN` right? The release process should be completely isolated, similar to how it was with Docker

File: make/buf/scripts/release.bash [hunk 6]
```diff
@@ -11,137 +11,42 @@ fail() {
...
-    mkdir -p "${dir}/share/zsh/site-functions"
-    #mkdir -p "${dir}/share/man/man1"
-    "$(uname -s)/$(uname -m)/${BASE_NAME}/bin/buf" bash-completion > "${dir}/etc/bash_completion.d/buf"
-    "$(uname -s)/$(uname -m)/${BASE_NAME}/bin/buf" fish-completion > "${dir}/share/fish/vendor_completions.d/buf.fish"
-    "$(uname -s)/$(uname -m)/${BASE_NAME}/bin/buf" zsh-completion > "${dir}/share/zsh/site-functions/_buf"
-    #"$(uname -s)/$(uname -m)/${1}/bin/buf" manpages "${dir}/share/man/man1"
-    cp -R "${DIR}/LICENSE" "${dir}/LICENSE"
-  done
-done
+echo "${RELEASE_MINISIGN_PRIVATE_KEY}" >"${WORKSPACE_DIR}/minisignsecret"
```
Comment (bufdev) [inline make/buf/scripts/release.bash:38]: These should not be echoed to a local file

File: make/buf/scripts/release.bash [hunk 7]
```diff
@@ -11,137 +11,42 @@ fail() {
...
-      arch="aarch64"
-    fi
-    dir="${os}/${arch}/${BASE_NAME}"
-    tar_context_dir="$(dirname "${dir}")"
-    tar_dir="${BASE_NAME}"
-    tarball="${BASE_NAME}-${os}-${arch}.tar.gz"
-    tar -C "${tar_context_dir}" -cvzf "${tarball}" "${tar_dir}"
-  done
-done
+mkdir -p "${WORKSPACE_DIR}/etc/bash_completion.d" \
```
Comment (bufdev) [inline make/buf/scripts/release.bash:41]: Just do separate `mkdir` invocations

File: make/buf/scripts/release.bash [hunk 8]
```diff
@@ -11,137 +11,42 @@ fail() {
...
+  "${WORKSPACE_DIR}/share/fish/vendor_completions.d" \
+  "${WORKSPACE_DIR}/share/zsh/site-functions"
 
-for file in $(find . -maxdepth 1 -type f | sed 's/^\.\///' | sort | uniq); do
-  sha256 "${file}" >> sha256.txt
-done
-sha256 -c sha256.txt
+"${GOBIN}/buf" bash-completion >"${WORKSPACE_DIR}/etc/bash_completion.d/buf"
```
Comment (bufdev) [inline make/buf/scripts/release.bash:45]: This should not use the project `buf` installation, it should use the release `buf` build

File: make/buf/scripts/release.bash [hunk 9]
```diff
@@ -11,137 +11,42 @@ fail() {
...
+"${GOBIN}/buf" fish-completion >"${WORKSPACE_DIR}/share/fish/vendor_completions.d/buf.fish"
+"${GOBIN}/buf" zsh-completion >"${WORKSPACE_DIR}/share/zsh/site-functions/_buf"
 
-mkdir -p assets
-for file in $(find . -maxdepth 1 -type f | sed 's/^\.\///' | sort | uniq); do
-  mv "${file}" "assets/${file}"
-done
+"${RELEASE_DIR}/go" mod verify
+env -i PATH="$PATH" HOME="$HOME" GOMODCACHE="$GOMODCACHE" GOCACHE="$GOCACHE" "$CACHE/bin/goreleaser" release ${SKIP_SIGN} ${SKIP_VALIDATE}
```
Comment (bufdev) [inline make/buf/scripts/release.bash:50]: What is `env -i`?

**Review by bufdev — COMMENTED** [review]
I'm still concerned about isolation here, added some comments.

**Comment by bufdev** [comment]
I'm not sure what we want to do with this - I think it'd be good to speed up the release process, but I also want to get this across the line or not do it at all. @amckinney we should talk about dedicating 2 eng hours to this and just get it across if we want to.

**Comment by akshayjshah** [comment]
It'll take longer than 2h, but I can look at this in conjunction with the build stuff for connect. If Alex has irreplaceable context, I can sync up with him and increase the bus factor a bit.

**Comment by bufdev** [comment]
Closing this for now - we can revisit in the future.

**Pull request closed without merging**